### PR TITLE
V4.0.0 - MultiStorage upgrade!

### PR DESCRIPTION
**Pull request by orsenkucher**
## Status
**IN DEVELOPMENT**

## Breaking Changes
NO

### MultiStorage upgrade!
* In-memory `Temporal` storage
* File-per-Bloc storage, by `Duplex` of default `cache` and `CellMultiplexer`
* All to single file storage, by `Duplex` of default `cache` and `CellSinglet`

### AES Encryption
* Use `StringCell` to save plain data
* Use `AESCell`(`BinaryCell`) to save encrypted binary
* Use `AESCell`(`Base64Cell`(`StringCell`)) to save encrypted base64 encoded text

### Safe Mode!
* Use `safeSync` to prevent data loss on switching multi-storage modes
* Non-breaking upgrade from v3.1.0 to v4.0.0

## Description-1
I have 2 branches on my repo.
`break-test-v3` with hydrated bloc v3.1.0
`break-test-v4` with hydrated bloc v4.0.0
It is used to ensure data compatibility between versions

## Description-2
The Plan B is to comment out interface for AES encryption, and release it with v4.1.0.
The Plan A is not to comment out code and fix aes/plain compatibility in this release.

## Description-3: Benchmark included!
```bash
$ cd benchmark
$ flutter run --release
```
Clear **benefits** of having different storages

## Description-4

This is merge of mine [v4.0.0](https://github.com/orsenkucher/hydrated_bloc/tree/v4.0.0) into your's [v4.0.0](https://github.com/felangel/hydrated_bloc/tree/v4.0.0)

### CODE REVIEW

**Comment by felangel** [comment]
@orsenkucher I took a quick look over the changes and I am wondering based on the benchmarks which is the most performant solution so far? Can you share the result of the benchmark?

I think we should carefully select the default storage mechanism and bundle that with the package and move other `HydratedStorage` implementations to separate package(s). Let me know what you think 👍

**Comment by orsenkucher** [comment]
@felangel 
Currently single file storage is needed to be v3.1.0 compatible, but multi-file does significantly better writes and is magnitudes better for encrypted writes

File: lib/src/hydrated_bloc.dart [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import 'dart:async';
-import 'dart:convert';
 
 import 'package:meta/meta.dart';
+import 'package:flutter/foundation.dart';
```
Comment (felangel) [inline lib/src/hydrated_bloc.dart]: what is this import needed for?

Comment (orsenkucher) [reply lib/src/hydrated_bloc.dart:4]: `freezed` uses it to make nicely readable objects in Flutter's devtools
This import is suggested in freezed docs

Comment (orsenkucher) [reply lib/src/hydrated_bloc.dart]: just some random import lol

**Comment by felangel** [comment]
@orsenkucher can we close this PR now that we've merged #52?

**Comment by orsenkucher** [comment]
💯 Yes! @felangel

**Pull request closed without merging**